flink-python/pyflink/fn_execution/boot.py

```diff
@@ -0,0 +1,352 @@
+        os.chmod(file_path, permissions)
+
+acceptable_whl_specs = [".whl"]
+
+
+def find_beam_sdk_whl():
+    """
+    Find apache beam sdk wheel package in downloaded files. If found, it will be installed.
+
+    :return: the file name of apache beam sdk wheel package.
+    """
+    for filename in files:
+        if filename.startswith("apache_beam"):
+            for acceptable_whl_spec in acceptable_whl_specs:
+                if filename.endswith(acceptable_whl_spec):
+                    logging.info("Found Apache Beam SDK wheel: %s" % filename)
+                    return filename
+    return ""
+
+
+def python_location():
+    """
+    Returns the python executable file location.
+    If the environment variable "python" exists, using its value.
+
+    :return: The python executable file location. Default value is "python".
+    """
+    if "python" in os.environ:
+        return os.environ["python"]
+    else:
+        return "python"
+
+
+def pip_command():
+    """
+    Returns the pip command.
+
+    If the environment variable "pip" exists, using its value.
+
+    :return: The pip executable file location. Default value is "/usr/local/bin/pip".
+    """
+    if "pip" in os.environ:
+        return [os.environ["pip"]]
+    else:
+        return [python_location(), "-m", "pip"]
+
+
+pip = pip_command()
+python_path = python_location()
+
+
+def add_test_params(args):
+    if "FLINK_BOOT_TESTING" in os.environ and os.environ["FLINK_BOOT_TESTING"] == "1":
+        # do not install those test packages in users' site-packages dir when testing
+        args.append("--prefix")
+        args.append(semi_persist_dir)
+        args.append("-I")
+    return args
+
+
+def pip_install_package(files, dir, name, force, optional, extras):
+    """
+    Pip install the specified python package.
+
+    :param files: The list of the downloaded file names.
+    :type files: list[str]
+    :param dir:  The path of the python package directory.
+    :type dir: str
+    :param name:  The file name of the python package.
+    :type name: str
+    :param force: Force install regardless of whether the package already installed.
+    :type force: bool
+    :param optional: If the specified python package not exists and this param is True,
+                     return 0. If the specified python package not exists and this param
+                     is False, raise an Exception.
+    :type optional: bool
+    :param extras: Specifies the extras_requires param of pip.
+    :type extras: list[str]
+    :return: 0 if install success or no need to install.
+    """
+    for filename in files:
+        if filename == name:
+            package = name
+            if extras is not None and len(extras) > 0:
+                package += "[" + ",".join(extras) + "]"
+            if force:
+                args = ["install", "--upgrade", "--force-reinstall", "--no-deps",
+                        os.path.join(dir, package)]
+                args[0:0] = pip
```

Comment:
refactor the code to remove the code duplication.

Reply:
Removed.